### PR TITLE
[SiliconFlow] Add _scaled_mm operator

### PR DESCRIPTION
**Pull request by winfan1314**
### PR Category
Operator / OP Test / Benchmark

### Type of Change
New Feature / Performance Optimization

### Description
This PR adds FlagGems support for `torch._scaled_mm` and `torch._scaled_mm.out`.

Changes included:
- Add `src/flag_gems/ops/scaled_mm.py` with a Triton implementation for scaled matrix multiplication.
- Support scalar and per-row/per-column scaling, optional bias, `out_dtype`, and the `out` variant.
- Add a CUDA CUTLASS fast path for supported FP8/INT8 SM90-SM99 cases, with Triton fallback for other supported floating dtypes/shapes.
- Add an Ascend large-aligned Triton fast path and backend tune configs, borrowing the tiling direction from existing Ascend `mm` patterns.
- Register `_scaled_mm` / `_scaled_mm.out` in `conf/operators.yaml` and FlagGems Python exports.
- Add accuracy coverage in `tests/test_scaled_mm.py` and benchmark coverage in `benchmark/test_scaled_mm.py`.

### Issue

Related to https://github.com/flagos-ai/FlagGems/issues/3387

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

#### Overall
The benchmark now follows the native-baseline policy from review feedback:

- CUDA SM90+ (for example H20): use native PyTorch target ops directly (`torch._scaled_mm` / `aten._scaled_mm.out`) and benchmark only PyTorch-supported FP8 (`float8_e4m3fn`) cases.
- CUDA pre-SM90 (for example A100/SM80): no supported native FP8 `_scaled_mm` benchmark case is available, so the benchmark skips cleanly instead of using a wrapper baseline.
- Ascend: `_scaled_mm` accuracy and benchmark tests are skipped with `flag_gems.vendor_name in ["ascend"]` because native PyTorch `_scaled_mm` / `aten._scaled_mm.out` support is unavailable there. The skip reflects native baseline/support limitations on this platform rather than CUDA FP8 benchmark behavior.

On NVIDIA H20 with `torch 2.9.0+cu128`, native PyTorch `_scaled_mm` accepts `float8_e4m3fn` input. `float16`, `float32`, and `bfloat16` inputs fail in native PyTorch with `Invalid scaling configuration`, and `float8_e5m2` fails with `Multiplication of two Float8_e5m2 matrices is not supported`.

#### Benchmark summary on NVIDIA H20 GPU
- command: `CUDA_VISIBLE_DEVICES=2 pytest -s benchmark/test_scaled_mm.py --level core --record log --tb=short`
- environment: `torch 2.9.0+cu128`, NVIDIA H20, SM90
- result: `8 passed in 29.18s`

| op | case | 16³ speedup | 128³ speedup | 512³ speedup |
|---|---|---:|---:|---:|
| `_scaled_mm` | fp8 scalar scale, fp16 output, bias | 0.967x | 0.965x | 1.008x |
| `_scaled_mm` | fp8 scalar scale, bf16 output, bias | 0.963x | 0.974x | 1.011x |
| `_scaled_mm` | fp8 scalar scale, fp32 output, no bias | 0.971x | 0.955x | 1.023x |
| `_scaled_mm` | fp8 row-wise scale, bf16 output, bias | 1.401x | 1.417x | 1.257x |
| `_scaled_mm.out` | fp8 scalar scale, fp16 output, bias | 0.944x | 0.978x | 1.011x |
| `_scaled_mm.out` | fp8 scalar scale, bf16 output, bias | 0.941x | 0.953x | 1.011x |
| `_scaled_mm.out` | fp8 scalar scale, fp32 output, no bias | 0.970x | 0.977x | 1.000x |
| `_scaled_mm.out` | fp8 row-wise scale, bf16 output, bias | 1.407x | 1.392x | 1.219x |

#### Skip-path validation
- Simulated CUDA pre-SM90/no-FP8 path: `2 skipped in 0.05s`
- Ascend accuracy + benchmark path: `16 skipped, 4 warnings in 0.23s`

### CODE REVIEW

File: benchmark/test_scaled_mm.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    def __init__(self):
+        dtypes = [torch.float16, torch.float32]
+        if flag_gems.runtime.device.support_bf16:
+            dtypes.append(torch.bfloat16)
+        if _cuda_fp8_available():
+            dtypes.append(torch.float8_e4m3fn)
+
+        super().__init__("scaled_mm", torch_scaled_mm_baseline, dtypes=dtypes)
+        self.set_gems(flag_gems.scaled_mm)
```
Comment (tengqm) [inline benchmark/test_scaled_mm.py]: Please avoid initializing the benchmark this way.
Pass all parameters from the test case entry please.

File: conf/operators.yaml [hunk 2]
```diff
@@ -4851,6 +4851,28 @@ ops:
...
+      multiplied by `scale_a` and `scale_b`, then an optional bias is added.
+    for:
+      - _scaled_mm
+    labels:
+      - aten
+    kind:
+      - BLAS
+    stages:
+      - beta: '5.1'
+  - id: scaled_mm_out
```
Comment (tengqm) [inline conf/operators.yaml:5358]: This has to be tested and benchmarked as well.

File: benchmark/test_scaled_mm.py [hunk 3]
```diff
@@ -62,18 +75,52 @@ def torch_scaled_mm_baseline(
...
-            dtypes.append(torch.bfloat16)
-        if _cuda_fp8_available():
-            dtypes.append(torch.float8_e4m3fn)
-
-        super().__init__("scaled_mm", torch_scaled_mm_baseline, dtypes=dtypes)
-        self.set_gems(flag_gems.scaled_mm)
+    def __init__(self, op_name, torch_op, gems_op, dtype, use_out=False):
+        super().__init__(op_name, torch_op, dtypes=[dtype])
+        self.set_gems(gems_op)
+        self.use_out = use_out
```
Comment (winfan1314) [inline benchmark/test_scaled_mm.py]: fixed `__init__`.

File: benchmark/test_scaled_mm.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    )
+    bench.run()
+
+
+@pytest.mark.scaled_mm_out
+@pytest.mark.parametrize("dtype", _benchmark_dtypes(), ids=_dtype_id)
+def test_scaled_mm_out_benchmark(dtype):
+    bench = ScaledMMBenchmark(
+        "scaled_mm_out",
+        torch_scaled_mm_out_baseline,
```
Comment (tengqm) [inline benchmark/test_scaled_mm.py]: can we avoid using wrappers for the target torch op?

Comment (winfan1314) [reply benchmark/test_scaled_mm.py]: Fixed, but now native `torch._scaled_mm` operator only supports fp8 on CUDA device. That's why I used a wrapper to add fp16, bf16 and fp32 support.

Comment (winfan1314) [reply benchmark/test_scaled_mm.py]: And more importantly, Ascend doesn't even have a backend native operator.

Comment (winfan1314) [reply benchmark/test_scaled_mm.py]: I think Ascend CI won't happy with this.

**Comment by winfan1314** [comment]
@tengqm Updated the benchmark split:

- CUDA SM90+ (e.g. H20): use native PyTorch target ops directly (`torch._scaled_mm` / `aten._scaled_mm.out`) and only benchmark PyTorch-supported FP8 (`float8_e4m3fn`) cases.
- CUDA pre-SM90 (e.g. A100/SM80): no supported native `_scaled_mm` FP8 benchmark case is available, so these cases are skipped instead of using a wrapper baseline.
- Ascend: keep performance coverage with an explicit reference baseline (`torch.mm + scale + bias + cast`) because native `torch._scaled_mm` / `aten._scaled_mm.out` is not available/supported there. No `torch_npu` private API is used.

Note: The FlagGems operator itself is still callable on the supported device/dtype combinations, including fp32/fp16/bf16 on CUDA and Ascend, plus fp8 on CUDA. The benchmark skips/restrictions are only due to limited native baseline support, not because the FlagGems implementation lacks those dtype paths.

File: benchmark/test_scaled_mm.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+
+@pytest.mark.scaled_mm_out
+@pytest.mark.parametrize("case", _benchmark_case_params())
+def test_scaled_mm_out_benchmark(case):
+    torch_op = (
+        torch_scaled_mm_out_reference
+        if _ascend_available()
+        else torch.ops.aten._scaled_mm.out
+    )
```
Comment (tengqm) [inline benchmark/test_scaled_mm.py]: My suggestion would be simply skip this test case if vendor_name is "ascend" (may need to add other vendors later). The philosophy is that we are not supposed to tweak the test case(s) to beautify the test results. If something doesn't work on a platform, it won't work even if that thing passed the tests. We leave this to the vendor to resolve. I mean a vendor is supposed to come up with a workaround, i.e. a customized implementation to make the operator work. Before that, we mark the test case with a `@pytest.mark.skip(vendor in ["ascend"], reason="Issue #1234: not working")` so that the balls are not dropped.

Thoughts?

Comment (winfan1314) [reply benchmark/test_scaled_mm.py]: Got it.

**Comment by bin913** [comment]
@winfan1314 Please fix the failing checks

**Comment by winfan1314** [comment]
@bin913 fixed ^-^

File: benchmark/test_scaled_mm.py (hunk 4 above)
Comment (tengqm) [reply benchmark/test_scaled_mm.py]: In that case, the protocol is to mark the test case with:

```
@pytest.mark.skip(vendor_name="ascend", reason="Issue #4567: Ascend isn't happy")
```

In other words, we file a known issue for this operator on Ascend and move forward.
The issue link makes sure that we are not pretending that it is working on Ascend and the issue has to be resolved later in this way or that.

File: benchmark/test_scaled_mm.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
+import pytest
+import torch
+
+import flag_gems
+
+from . import base, consts
+
+pytestmark = pytest.mark.skipif(
+    flag_gems.vendor_name in ["ascend"],
+    reason="Native torch._scaled_mm is not supported on Ascend.",
```
Comment (tengqm) [inline benchmark/test_scaled_mm.py]: This warrants a dedicated issue.
We don't skip test cases without issues filed.
A test case can be skipped, but a skip is has to be a known defect that can be tracked publicly.

**Review by tengqm — COMMENTED** [review]
I'm okay with the change, provided that we file an issue and change the skip reason to a link to that issue.
The important thing is that we don't get balls dropped.

File: benchmark/test_scaled_mm.py (hunk 6 above)
Comment (winfan1314) [reply benchmark/test_scaled_mm.py]: OK I'll draft an issue.

File: benchmark/test_scaled_mm.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
+import pytest
+import torch
+
+import flag_gems
+
+from . import base, consts
+
+pytestmark = pytest.mark.skipif(
+    flag_gems.vendor_name in ["ascend"],
+    reason="https://github.com/flagos-ai/FlagGems/issues/3387",
```
Comment (tengqm) [inline benchmark/test_scaled_mm.py]:
```suggestion
    reason="Issue #3387: Not supported on Ascend",
```

**Review by tengqm — DISMISSED** [review]

File: tests/test_scaled_mm.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        return scale
+    if scale.ndim == 1:
+        if is_left_scale and scale.shape[0] == rows:
+            return scale.reshape(rows, 1)
+        if not is_left_scale and scale.shape[0] == cols:
+            return scale.reshape(1, cols)
+    return scale
+
+
+def _reference_scaled_mm(mat1, mat2, scale_a, scale_b, bias, out_dtype):
```
Comment (w1120029931-bit) [inline tests/test_scaled_mm.py:97]: **Issue**: Reference computation does not use utils.to_reference(). The helper currently converts tensors with detach().cpu().float() directly.

**Recommended fix:**
Use utils.to_reference() for tensors involved in the reference computation, keeping the reference path consistent with the project test convention.

<sub>Rule: `TEST_TO_REFERENCE`</sub>

**Review by w1120029931-bit — COMMENTED** [review]
## FlagGems Automated Review

**Operator:** `scaled_mm`
**Verdict:** COMMENT

### Summary
- Errors: 1
- Warnings: 0
- Suggestions: 0

**Review by w1120029931-bit — APPROVED** [review]
